File: stars.cc
#include <stdio.h>
#include <minifb/MiniFB.h>
#include <array>
#include <algorithm>
#include <stdlib.h>
#include <time.h>
#include <limits>

auto title = "Star Field CC";
constexpr auto width = 800;
constexpr auto height = 600;
constexpr auto half_width = width / 2;
constexpr auto half_height = height / 2;
auto pixels = std::array<unsigned int, width * height>{0};
constexpr auto white = (255 << 16) | (255 << 8) | 255;
auto clear = []{ std::fill(pixels.begin(), pixels.end(), 0); };

constexpr auto num_stars = 1000;
auto spread = 20;
auto speed  = 0.1;

float star_x[num_stars]{0};
float star_y[num_stars]{0};
float star_z[num_stars]{0};

auto random_float = []{ return (double)rand() / (double)RAND_MAX; };

void init_star(int i) {
    star_x[i] = 2 * (random_float() - .5) * spread; 
    star_y[i] = 2 * (random_float() - .5) * spread; 
    star_z[i] = (random_float() + .00001) * spread;
}

int main() {
    srand(time(0));

    for (auto i = 0; i < num_stars; i++) init_star(i);

    auto window = mfb_open(title, width, height);

    while (mfb_wait_sync(window)) {
        if (mfb_update_ex(window, pixels.data(), width, height) < 0 ) break;
        // clear background
        clear(); 
        // draw stars
        for (auto i = 0; i < num_stars; i++) {
            auto& z = star_z[i];
            z -= speed;
             
            if ( z <= 0 ) init_star(i);

            // convert to screen space
            int x = (star_x[i] / z) * half_width + half_width;
            int y = (star_y[i] / z) * half_height + half_height;
            
            if ( x < 0 or x >= width or y < 0 or y >= height )
                init_star(i);
            else
                pixels[ y * width + x ] = white;
        } 
    }
}

File: stars.nim
import minifb,
       minifb_enums,
       std/algorithm, # fill
       std/random

const
    title = "Star Field Nim"
    width = 800
    height = 600
    half_width = width / 2
    half_height = height / 2
    white = (255 shl 16) or (255 shl 8) or 255

    num_stars = 1000
    spread = 20
    speed = 0.1

type
    Star_Range = array[ num_stars, float ]
var
    pixels : array[ width * height, uint32 ]
    star_x : Star_Range
    star_y : Star_Range
    star_z : Star_Range

proc clear() = fill(pixels, 0)

proc init_star(i: int) =
    star_x[i] = 2 * (rand(1.0) - 0.5) * spread
    star_y[i] = 2 * (rand(1.0) - 0.5) * spread
    star_z[i] = (rand(1.0) + 0.00001) * spread

randomize()

for i in 0 .. num_stars - 1: init_star(i)

let window = mfb_open(title.cstring, width.cuint, height.cuint)

while mfb_wait_sync(window):
    if mfb_update_ex(window, pixels.addr, width.cuint, height.cuint) != STATE_OK: break

    clear()

    for i in 0 .. num_stars - 1:
        let z = star_z[i].addr
        z[] -= speed

        if z[] <= 0: init_star(i)

        let x = int((star_x[i] / z[]) * half_width + half_width)
        let y = int((star_y[i] / z[]) * half_height + half_height)

        if x < 0 or x >= width or y < 0 or y >= height:
            init_star(i)
        else:
            pixels[ y * width + x ] = white
        
